[inductor] remove reordering_reindex

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127367

This fixes the loop ordering issue for avg_pool2d here (https://github.com/pytorch/pytorch/issues/126255#issuecomment-2117931529).

The reason we can not fuse the 2 kernels for avg_pool2d is due to ComputedBuffer.iter_reordering_reindex. Take a simpler example:

```
        def f(x, y):
            """
            Add a matmul since inductor may force layout for output.
            """
            return (x.sum(dim=-1) + 1) @ y

        # Make the first 2 dimension not able to merge on purpose so that
        # ComputedBuffer.iter_reoredering_reindex will be updated.
        x = rand_strided([20, 20, 30], [30, 900, 1], device="cuda")
        y = torch.randn(20, 20)
```

Suppose x.sum is stored to x2. The computed buffer for x2 will remember that we have reordered it's first and second dimension (i.e. loop order [1, 0]). Later one when we decide the loop order for x2 when computing 'x2 + 1' , we decide to pick loop order [1, 0] according to the stride analysis. And then we use the saved ComputedBuffer.iter_reordering_reindex to further reorder the loop order. The net effect is that we use loop order [0, 1] which cause the pointwise kernel not able to fuse with the reduction kernel.

I feel that we don't need ComputedBuffer.iter_reordering_reindex. And test result shows removing it has neutral impact on the dashboard [link](https://hud.pytorch.org/benchmark/compilers?startTime=Wed%2C%2022%20May%202024%2017%3A30%3A29%20GMT&stopTime=Wed%2C%2029%20May%202024%2017%3A30%3A29%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/153/head&lCommit=195f42cf1a414d2d1a0422b8a081a85ff52b7d20&rBranch=main&rCommit=d6e3e89804c4063827ea21ffcd3d865e5fe365d9)



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

shunting314 committed:
[temp] ignore reordering_reindex

[ghstack-poisoned]

shunting314 committed:
Update on "[temp] ignore reordering_reindex"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee reviewed (commented):
For these kinds of tests I like checking memory reads + writes. That's always just in `metrics.num_bytes_accessed` now.

eellison commented on torch/_inductor/ir.py:
```diff
@@ -3435,16 +3427,15 @@ def simplify_and_reorder(x_vars, support_vars, sizes, reordering_reindex=None):
             reduce_vars, support_vars, reduce_size
         )
 
-        # remember the reordering if not have loop collapse.
-        if len(iter_ranges) == len(index_vars):
-            self.iter_reordering_reindex = iter_reordering_reindex
```
Nice - reordering_reindex had caused issues previously in https://github.com/pytorch/pytorch/pull/96640/files#diff-cf6ca00beddc32a2a6a2933fb9913b6a2b925ffc3b745488967210e4343134acR2237.

shunting314 commented:
mobilenetv3_large_100 accuracy test starts to fail. Since it's not real accuracy problem, I'll increase the tolerance. BTW, the accuracy test pass on the dashboard.

shunting314 committed:
Update on "[inductor] remove reordering_reindex"


This fixes the loop ordering issue for avg_pool2d here (https://github.com/pytorch/pytorch/issues/126255#issuecomment-2117931529).

The reason we can not fuse the 2 kernels for avg_pool2d is due to ComputedBuffer.iter_reordering_reindex. Take a simpler example:

```
        def f(x, y):
            """
            Add a matmul since inductor may force layout for output.
            """
            return (x.sum(dim=-1) + 1) @ y

        # Make the first 2 dimension not able to merge on purpose so that
        # ComputedBuffer.iter_reoredering_reindex will be updated.
        x = rand_strided([20, 20, 30], [30, 900, 1], device="cuda")
        y = torch.randn(20, 20)
```

Suppose x.sum is stored to x2. The computed buffer for x2 will remember that we have reordered it's first and second dimension (i.e. loop order [1, 0]). Later one when we decide the loop order for x2 when computing 'x2 + 1' , we decide to pick loop order [1, 0] according to the stride analysis. And then we use the saved ComputedBuffer.iter_reordering_reindex to further reorder the loop order. The net effect is that we use loop order [0, 1] which cause the pointwise kernel not able to fuse with the reduction kernel.

I feel that we don't need ComputedBuffer.iter_reordering_reindex. And test result shows removing it has neutral impact on the dashboard [link](https://hud.pytorch.org/benchmark/compilers?startTime=Wed%2C%2022%20May%202024%2017%3A30%3A29%20GMT&stopTime=Wed%2C%2029%20May%202024%2017%3A30%3A29%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/153/head&lCommit=195f42cf1a414d2d1a0422b8a081a85ff52b7d20&rBranch=main&rCommit=d6e3e89804c4063827ea21ffcd3d865e5fe365d9)



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

shunting314 commented:
@pytorchbot merge

nWEIdia commented:
Is this causing deterministic accuracy failure for cspdarknet? 


![image](https://github.com/pytorch/pytorch/assets/143543872/f9217ef9-e286-4c40-9c5c-e528569dbef2)

https://hud.pytorch.org/hud/pytorch/pytorch/main/2?per_page=50&name_filter=inductor_timm

It is a bit unlike cu124 gluon_inception failure, where most of the time it was "improving accuracy"

cc @huydhn @clee2000 @atalman

shunting314 commented:
@nWEIdia the UI shows the cspdarknet53 is flaky on trunk though.

<img width="822" alt="Screenshot 2024-06-03 at 5 00 46 PM" src="https://github.com/pytorch/pytorch/assets/52589240/de6b045e-5dc6-4f13-97eb-6b6202210ed8">

nWEIdia commented:
> @nWEIdia the UI shows the cspdarknet53 is flaky on trunk though.
> 
> <img alt="Screenshot 2024-06-03 at 5 00 46 PM" width="822" src="https://private-user-images.githubusercontent.com/52589240/336260673-de6b045e-5dc6-4f13-97eb-6b6202210ed8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc0NjIyNTksIm5iZiI6MTcxNzQ2MTk1OSwicGF0aCI6Ii81MjU4OTI0MC8zMzYyNjA2NzMtZGU2YjA0NWUtNWRjNi00ZjEzLTk3ZWItNmI2MjAyMjEwZWQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA0VDAwNDU1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk2NGZmNjA1ODllMTZkNzBjMzlkYTg3ZDUwZDBhNWY3YmMyYTJhZmQ4OTkxMTlmM2U4YWNmN2FhMDViYzY5MDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.OqPRnAwpbTLV_sTqCxKkqCqLVZQBVK8lUlmLYPfc3mc">

The UI can be wrong. My understanding is before this PR, some other model was flakily failing. But after this PR, cspdarknent53 is consistently failing. 

You could go to hud.pytorch.org and use inductor_timm as filter and you could see the determinism. 

It is still possible that something else might be going on, but I would like to raise the non-flaky part of cspdarknet model. It is an oversight to say a model is flaky/unstable while it is stably failing.

cc @malfet @ptrblck

nWEIdia commented:
@shunting314 The revert PR made the cspdarknet model pass. See https://github.com/pytorch/pytorch/actions/runs/9359613196/job/25764625962#step:20:1683 

Should this PR be reverted along with the subsequent changes? https://github.com/pytorch/pytorch/pull/127807

shunting314 commented:
@nWEIdia here is a forward fix: https://github.com/pytorch/pytorch/pull/127949
